pkg/controller/master/upgradelog/common.go

```diff
@@ -0,0 +1,1008 @@
+package upgradelog
+
+import (
+	"errors"
+	"fmt"
+
+	loggingv1 "github.com/banzaicloud/logging-operator/pkg/sdk/logging/api/v1beta1"
+	"github.com/banzaicloud/logging-operator/pkg/sdk/logging/model/filter"
+	"github.com/banzaicloud/logging-operator/pkg/sdk/logging/model/output"
+	"github.com/banzaicloud/operator-tools/pkg/volume"
+	"github.com/rancher/fleet/pkg/apis/fleet.cattle.io/v1alpha1"
+	mgmtv3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	ctlcorev1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/rancher/wrangler/pkg/name"
+	appsv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	harvesterv1 "github.com/harvester/harvester/pkg/apis/harvesterhci.io/v1beta1"
+)
+
+const (
+	defaultJobBackoffLimit    int32 = 5
+	defaultDeploymentReplicas int32 = 1
+	logPackagingScript              = `
+#!/usr/bin/env sh
+set -e
+
+echo "start to package upgrade logs"
+
+archive="$ARCHIVE_NAME.tar.gz"
+tmpdir=$(mktemp -d)
+mkdir $tmpdir/logs
+
+cd /archive/logs
+
+for f in *.log
+do
+    cat $f | awk '{$1=$2=""; print $0}' | jq -r .message > $tmpdir/logs/$f
+done
+
+tar -zcvf /archive/$archive -C $tmpdir .
+ls -l /archive/$archive
+echo "done"
+`
+)
+
+const (
+	InfraComponent      = "infra"
+	ShipperComponent    = "shipper"
+	AggregatorComponent = "aggregator"
+	DownloaderComponent = "downloader"
+	FlowComponent       = "clusterflow"
+	LogArchiveComponent = "log-archive"
+	OperatorComponent   = "operator"
+	OutputComponent     = "clusteroutput"
+	PackagerComponent   = "packager"
+)
+
+func upgradeLogReference(upgradeLog *harvesterv1.UpgradeLog) metav1.OwnerReference {
+	return metav1.OwnerReference{
+		Name:       upgradeLog.Name,
+		Kind:       upgradeLog.Kind,
+		UID:        upgradeLog.UID,
+		APIVersion: upgradeLog.APIVersion,
+	}
+}
+
+func preparePvc(upgradeLog *harvesterv1.UpgradeLog) *corev1.PersistentVolumeClaim {
+	upgradeLogStorageClassName := harvesterUpgradeLogStorageClassName
+	volumeMode := harvesterUpgradeLogVolumeMode
+
+	return &corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels: map[string]string{
+				harvesterUpgradeLogLabel:          upgradeLog.Name,
+				harvesterUpgradeLogComponentLabel: LogArchiveComponent,
+			},
+			Name:      name.SafeConcatName(upgradeLog.Name, LogArchiveComponent),
+			Namespace: upgradeLogNamespace,
+			OwnerReferences: []metav1.OwnerReference{
+				upgradeLogReference(upgradeLog),
+			},
+		},
+		Spec: corev1.PersistentVolumeClaimSpec{
+			AccessModes: []corev1.PersistentVolumeAccessMode{
+				corev1.ReadWriteOnce,
```

Comment:
Yes, it will follow the fluentd Pod. And other Pods like the log packager and downloader have PodAffinities specified to coexist with the fluentd Pod.